modules/generator/registry/native_histogram.go

```diff
@@ -0,0 +1,336 @@
+		onAddSeries = func(uint32) bool {
+			return true
+		}
+	}
+	if onRemoveSeries == nil {
+		onRemoveSeries = func(uint32) {}
+	}
+
+	if traceIDLabelName == "" {
+		traceIDLabelName = "traceID"
+	}
+
+	return &nativeHistogram{
+		metricName:       name,
+		series:           make(map[uint64]*nativeHistogramSeries),
+		onAddSerie:       onAddSeries,
+		onRemoveSerie:    onRemoveSeries,
+		traceIDLabelName: traceIDLabelName,
+		buckets:          buckets,
+	}
+}
+
+func (h *nativeHistogram) ObserveWithExemplar(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) {
+	hash := labelValueCombo.getHash()
+
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	s, ok := h.series[hash]
+	if ok {
+		h.updateSeries(s, value, traceID, multiplier)
+		return
+	}
+
+	if !h.onAddSerie(h.activeSeriesPerHistogramSerie()) {
+		return
+	}
+
+	newSeries := h.newSeries(labelValueCombo, value, traceID, multiplier)
+	h.series[hash] = newSeries
+}
+
+func (h *nativeHistogram) newSeries(labelValueCombo *LabelValueCombo, value float64, traceID string, multiplier float64) *nativeHistogramSeries {
+	newSeries := &nativeHistogramSeries{
+		// TODO move these labels in HistogramOpts.ConstLabels?
+		labels: labelValueCombo.getLabelPair(),
+		promHistogram: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:    h.name(),
+			Help:    "Native histogram for metric " + h.name(),
+			Buckets: h.buckets,
+			// TODO check if these values are sensible and break them out
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 15 * time.Minute,
+		}),
+		lastUpdated: atomic.NewInt64(0),
+	}
+
+	h.updateSeries(newSeries, value, traceID, multiplier)
+
+	return newSeries
+}
+
+func (h *nativeHistogram) updateSeries(s *nativeHistogramSeries, value float64, traceID string, multiplier float64) {
+	for i := 0.0; i < multiplier; i++ {
+		s.promHistogram.(prometheus.ExemplarObserver).ObserveWithExemplar(
+			value,
+			map[string]string{h.traceIDLabelName: traceID},
+		)
+	}
+	s.lastUpdated.Store(time.Now().UnixMilli())
+}
+
+func (h *nativeHistogram) name() string {
+	return h.metricName
+}
+
+func (h *nativeHistogram) collectMetrics(appender storage.Appender, timeMs int64, externalLabels map[string]string) (activeSeries int, err error) {
+	h.seriesMtx.Lock()
+	defer h.seriesMtx.Unlock()
+
+	labelsCount := 0
+	if h.series[0] != nil {
+		labelsCount = len(h.series[0].labels.names)
+	}
+	lbls := make(labels.Labels, 1+len(externalLabels)+labelsCount)
+	lb := labels.NewBuilder(lbls)
+	activeSeries = 0
+
+	lb.Set(labels.MetricName, h.metricName)
+
+	// set external labels
+	for name, value := range externalLabels {
+		lb.Set(name, value)
+	}
+
+	for _, s := range h.series {
+
+		// Set series-specific labels
+		for i, name := range s.labels.names {
+			lb.Set(name, s.labels.values[i])
+		}
+
+		// Extract histogram
+		encodedMetric := &dto.Metric{}
+
+		// Encode to protobuf representation
```

Comment:
i don't think this encodes to protobuf (which is good. that would be costly). it just does a fairly complicated transition from one representation to another

Reply:
Its doing something protobuf.  Perhaps not a `Marshall` call, but   dto.Metric is a protobuf struct.  [Here](https://github.com/zalegrala/tempo/blob/3ff3eb85abb298739b44b42ab444632ab815b996/vendor/github.com/prometheus/client_golang/prometheus/histogram.go#L751) is the `Write` method.